stratum/hal/lib/barefoot/bf_chassis_manager.cc

```diff
@@ -351,13 +352,94 @@ ::util::Status BFChassisManager::PushChassisConfig(
+        << "for another Node in the config.";
+  }
+  {
+    int unit = 0;
+    for (const auto& node : config.nodes()) {
+      unit_to_node_id[unit] = node.id();
+      node_id_to_unit[node.id()] = unit;
+      ++unit;
+    }
   }
-  if (config.port_groups_size()) {
-    return MAKE_ERROR(ERR_INVALID_PARAM)
-           << "Port groups are not supported on Tofino.";
+
+  // Go over all the singleton ports in the config:
+  // 1- Validate the basic singleton port properties.
+  // 2- Make sure there is no two ports with the same (slot, port, channel).
+  // 3- Make sure for each (slot, port) pair, the channels of all the ports
+  //    are valid. This depends on the port speed.
+  // 4- Make sure no singleton port has the reserved CPU port ID. CPU port is
+  //    a special port and is not in the list of singleton ports. It is
+  //    configured separately.
+  // 5- Make sure IDs of the singleton ports are unique per node.
+  std::map<uint64, std::set<uint32>> node_id_to_port_ids;
+  std::set<PortKey> singleton_port_keys;
+  for (const auto& singleton_port : config.singleton_ports()) {
+    // TODO(max): enable once we decoupled port ids from sdk ports.
+    // CHECK_RETURN_IF_FALSE(singleton_port.id() > 0)
+    //     << "No positive ID in " << PrintSingletonPort(singleton_port) << ".";
+    CHECK_RETURN_IF_FALSE(singleton_port.id() != kCpuPortId)
+        << "SingletonPort " << PrintSingletonPort(singleton_port)
+        << " has the reserved CPU port ID (" << kCpuPortId << ").";
+    CHECK_RETURN_IF_FALSE(singleton_port.slot() > 0)
+        << "No valid slot in " << singleton_port.ShortDebugString() << ".";
+    CHECK_RETURN_IF_FALSE(singleton_port.port() > 0)
+        << "No valid port in " << singleton_port.ShortDebugString() << ".";
+    CHECK_RETURN_IF_FALSE(singleton_port.channel() == 0)
+        << "SingletonPort " << singleton_port.ShortDebugString()
+        << " contains unsupported channel field.";
+    CHECK_RETURN_IF_FALSE(singleton_port.speed_bps() > 0)
+        << "No valid speed_bps in " << singleton_port.ShortDebugString() << ".";
+    PortKey singleton_port_key(singleton_port.slot(), singleton_port.port(),
+                               singleton_port.channel());
+    CHECK_RETURN_IF_FALSE(!singleton_port_keys.count(singleton_port_key))
+        << "The (slot, port, channel) tuple for SingletonPort "
+        << PrintSingletonPort(singleton_port)
+        << " was already recorded for another SingletonPort in the config.";
+    singleton_port_keys.insert(singleton_port_key);
+    CHECK_RETURN_IF_FALSE(singleton_port.node() > 0)
+        << "No valid node ID in " << singleton_port.ShortDebugString() << ".";
+    CHECK_RETURN_IF_FALSE(node_id_to_unit.count(singleton_port.node()))
+        << "Node ID " << singleton_port.node() << " given for SingletonPort "
+        << PrintSingletonPort(singleton_port)
+        << " has not been given to any Node in the config.";
+    CHECK_RETURN_IF_FALSE(
```

Comment:
Does this mean that port IDs only need to be unique for a given node? Said another way, this check verifies that (port id, node id) is unique...

Comment in ChassisConfig message says that the port ID needs to be unique across all ports (including across other port types i.e. trunks).

Reply:
> this check verifies that (port id, node id) is unique

This is correct. The tuple (node, port id) must be unique per switch.

The code is taken nearly verbatim from the BCM chassis manager, but we could ask at our next sync.

To me this seems sensible. Given that the port ID is a configurable SDN port, why should it not be allows to have port 1 on both forwarding planes? P4RT requests always contain the node ID.